.github/workflows/unittests.yml

```diff
@@ -2,24 +2,39 @@ name: Tests
 
 on:
   push:
-    branches: [ main ]
+    branches: [main]
   pull_request:
-    branches: [ main ]
+    branches: [main]
+    types:
+      - opened
+      - ready_for_review
+      - reopened
+      - synchronize
   workflow_call:
 
 jobs:
   test_package:
     name: Unit Tests
     runs-on: "ubuntu-20.04"
     steps:
-        - uses: actions/checkout@v4
-        - name: Setup python
-          uses: actions/setup-python@v5
-          with:
-            python-version: "3.11"
-            architecture: x64
-        - name: Install dependencies
-          run: |
-            pip install -r code/requirements.txt -r code/dev-requirements.txt -r code/backend/requirements.txt
-        - name: Run Python tests
-          run: cd code/; python -m pytest -m "not azure"
+      - uses: actions/checkout@v4
+      - name: Setup python
+        uses: actions/setup-python@v5
+        with:
+          python-version: "3.11"
+          architecture: x64
+          cache: pip
+          cache-dependency-path: "**/*requirements*.txt"
+      - name: Install dependencies
+        run: |
+          pip install -r code/requirements.txt -r code/dev-requirements.txt -r code/backend/requirements.txt
+      - name: Run Python tests
+        run: python -m pytest -m "not azure" --junitxml=junit-coverage.xml --cov --cov-report xml:coverage.xml
+        working-directory: code
+      - uses: actions/upload-artifact@v4
```

Comment:
Where does this upload the coverage report too out of curiosity? How does it differentiate coverage reports from different workflow runs? Is there a retention period?

Sorry for all the questions, I've not used it before!

Reply:
The artifacts are tied to a specific workflow run, however if you have a second workflow using `workflow_run`, then you can still pull artifacts from the workflow that called it, which is whats's going on here.

There is a retention period for artifacts, you can either set your own value here or it will fall back to whatever the default is for the repository.